lib/active_model/serializer/type.rb

```diff
@@ -16,8 +16,13 @@ module ClassMethods
         # @example
         #   class AdminAuthorSerializer < ActiveModel::Serializer
         #     type 'authors'
+        #
+        # @example
+        # TODO: actually do block, not Proc
+        #   class AdminAuthorSerializer < ActiveModel::Serializer
+        #     type { |object| object.class.name }
         def type(type)
-          self._type = type && type.to_s
+          self._type = type.is_a?(Proc) ? type : type.to_s if type
```

Comment:
@bf4 why do you think it causes memory leak? I assume you would have a limited amount of serializer and therefore a limited amount of proc and contexts too.